backend/src/Edna/Upload/Service.hs

```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+
+      measurementsToSummary fileMeasurements
+
+insertTarget :: Text -> Edna (Text, Word32)
+insertTarget targetName = (targetName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esTarget ednaSchema)
+    (insertExpressions [TargetRec default_ (val_ targetName) default_])
+    (Pg.onConflict (Pg.conflictingFields tName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added target not found: " <> targetName)  <$>
+    targetNameToId targetName
+
+insertCompound :: Text -> Edna (Text, Word32)
+insertCompound compoundName = (compoundName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esCompound ednaSchema)
+    (insertExpressions [CompoundRec default_ (val_ compoundName) default_ default_])
+    (Pg.onConflict (Pg.conflictingFields cName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added compound not found: " <> compoundName)  <$>
+    compoundNameToId compoundName
+
+insertExperimentFile ::
+  Word32 -> Word32 -> FileMetadata -> Text -> Text -> LByteString -> Edna Word32
+insertExperimentFile projId methodId meta descr fileName blob = do
+  unSerial . expectOneInsertion . map efExperimentFileId <$> runInsertReturningList'
+    (insert (esExperimentFile ednaSchema) $ insertExpressions [experimentFileRec])
+  where
+    experimentFileRec :: ExperimentFileT (QExpr Postgres s)
+    experimentFileRec = ExperimentFileRec
+      { efExperimentFileId = default_
+      , efProjectId = val_ projId
+      , efMethodologyId = val_ (Just methodId)
+      , efUploadDate = default_
+      , efMeta = val_ (PgJSON meta)
+      , efDescription = val_ descr
+      , efName = val_ fileName
+      , efContents = val_ blob
+      }
+
+insertExperiment :: Word32 -> Word32 -> Word32 -> [Measurement] -> Edna ()
+insertExperiment experimentFileId compoundId targetId measurements = do
+  expId <-
+    unSerial . expectOneInsertion . map eExperimentId <$> runInsertReturningList'
+    (insert (esExperiment ednaSchema) $ insertExpressions [experimentRec])
+  subExpId <-
+    unSerial . expectOneInsertion . map seSubExperimentId <$> runInsertReturningList'
+    (insert (esSubExperiment ednaSchema) $ insertExpressions [subExperimentRec expId])
+  forM_ measurements $ \measurement -> do
+    let
+      measurementRec :: MeasurementT (QExpr Postgres s)
+      measurementRec = MeasurementRec
+        { mMeasurementId = default_
+        , mExperimentId = val_ expId
+        , mConcentration = val_ (EReader.mConcentration measurement)
+        , mSignal = val_ (EReader.mSignal measurement)
+        , mIsOutlier = val_ (EReader.mIsOutlier measurement)
+        }
+    measurementId <-
+      unSerial . expectOneInsertion . map mMeasurementId <$> runInsertReturningList'
+      (insert (esMeasurement ednaSchema) $ insertExpressions [measurementRec])
+    when (EReader.mIsOutlier measurement) $
+      runInsert' $ insert (esRemovedMeasurements ednaSchema) $ insertValues
+      [RemovedMeasurementsRec subExpId measurementId]
```

Comment:
I am not sure that I am right, but as I understand call sql `1` time is better than calling it `n` times. So, you can construct list of `measurementRec` and insert them using `runInsertReturningList` and get it's `ids` with `isOutliers`, filter on `isOutliers` and again construct a list of sub-experiments and simply call sql one time

Reply:
Ok, sounds doable. I suppose there will be 2 SQL calls instead of this `forM_`, not 1.